ePBFT/src/pbft/libbyz/LedgerReplay.cpp

```diff
@@ -36,54 +27,70 @@ std::unique_ptr<Pre_prepare> LedgerReplay::process_data(
   PBFT_ASSERT(
     !data.empty(), "apply ledger data should not receive empty vector");
 
-  // received a new entry so we have read the entry size from the file
-  total_ledger_len += sizeof(size_t);
-
   auto entry_data = data.data();
   auto data_size = data.size();
-  total_ledger_len += data_size;
 
-  // peek at header type
-  Ledger_header_type type =
-    *reinterpret_cast<Ledger_header_type*>(const_cast<uint8_t*>(entry_data));
+  std::vector<std::unique_ptr<Pre_prepare>> pre_prepares;
 
-  if (type == Ledger_header_type::Pre_prepare_ledger_header)
+  while (data_size > 0)
   {
-    auto header =
-      serialized::overlay<Pre_prepare_ledger_header>(entry_data, data_size);
-    latest_pre_prepare = create_message<Pre_prepare>(entry_data, data_size);
-  }
-  else if (type == Ledger_header_type::Pre_prepare_ledger_large_message_header)
-  {
-    auto header = serialized::overlay<Pre_prepare_ledger_large_message_header>(
-      entry_data, data_size);
-
-    auto req =
-      create_message<Request>(entry_data, header.message_size).release();
+    serialized::skip(entry_data, data_size, sizeof(uint32_t));
+    // peek at header type
+    Ledger_header_type type =
+      *reinterpret_cast<Ledger_header_type*>(const_cast<uint8_t*>(entry_data));
 
-    if (!(req->size() > Request::big_req_thresh && brt.add_request(req)))
+    if (type == Ledger_header_type::Pre_prepare_ledger_header)
     {
-      rqueue.append(req);
+      auto header =
+        serialized::overlay<Pre_prepare_ledger_header>(entry_data, data_size);
+      latest_pre_prepare =
+        create_message<Pre_prepare>(entry_data, header.message_size);
+      serialized::skip(entry_data, data_size, header.message_size);
+      if (latest_pre_prepare->num_big_reqs() > 0)
+      {
+        for (size_t i = 0; i < latest_pre_prepare->num_big_reqs(); ++i)
+        {
+          auto header =
+            serialized::overlay<Pre_prepare_ledger_large_message_header>(
+              entry_data, data_size);
+
+          auto req =
+            create_message<Request>(entry_data, header.message_size).release();
+
+          serialized::skip(entry_data, data_size, header.message_size);
+          if (!(req->size() > Request::big_req_thresh && brt.add_request(req)))
+          {
+            rqueue.append(req);
+          }
+        }
+      }
     }
-  }
 
-  auto pre_prepare = latest_pre_prepare.get();
+    auto pre_prepare = latest_pre_prepare.get();
 
-  Pre_prepare::Requests_iter iter(pre_prepare);
-  // check that all big requests are present
-  Request request;
-  bool is_request_present = false;
-  while (iter.get_big_request(request, is_request_present))
-  {
-    if (!is_request_present)
+    Pre_prepare::Requests_iter iter(pre_prepare);
+    // check that all big requests are present
+    Request request;
+    bool is_request_present = false;
```

Comment:
I'm not comfortable introducing new code to accomodate big request. Can we please ticket the removal of this logic?

Reply:
Big requests is the way that we have been dealing with all requests in PBFT. It works. Removing it will add work and remove an optimization that offloads request broadcast from the primary. Why do you want to remove it?

Reply:
As discussed, let's just call them requests from now and forget about inline requests.